### PR TITLE
mock event processor client live tests

### PR DESCRIPTION
**Pull request by christothes**
PR to address #7541

**Summary of changes**:
- creation of `MockEventProcessorClient`
  - overrides `LoadBalanceUpdate` and `OwnershipExpiration`
  - tracks calls to all event delegates for validation
    - `ProcessExceptionAsync`
    - `ProcessEventAsync`
    - `InitializeProcessingForPartitionAsync`
    - `ProcessingForPartitionStoppedAsync`
  - implements `WaitStabilization` from `EventProcessorManager`
- Moved non-failing and stable tests from `EventProcessorClientLiveTests` to `EventProcessorClientTests`
  - some tests were not removed from `EventProcessorClientLiveTests` if they appeard to have integration test value.

**Test Coverage Diff for EventProcessorClient:**
Before:
```
+---------+--------+--------+--------+
|         | Line   | Branch | Method |
+---------+--------+--------+--------+
| Total   | 49.73% | 53.61% | 59.01% |
+---------+--------+--------+--------+
| Average | 74.86% | 76.8%  | 79.5%  |
+---------+--------+--------+--------+
```

After:
```
+---------+--------+--------+--------+
|         | Line   | Branch | Method |
+---------+--------+--------+--------+
| Total   | 79.3%  | 82.42% | 78.68% |
+---------+--------+--------+--------+
| Average | 89.65% | 91.21% | 89.34% |
+---------+--------+--------+--------+
```

### CODE REVIEW

**Comment by jsquire** [comment]
Thanks very much for your contribution, @christothes.   Apologies for the delay in reviewing; this week has been a bit busy.   I'll do my best to carve out time tomorrow to give this a good look, though there's a fair chance that I may not be able to do so until Monday.

**Comment by jsquire** [comment]
@kinelski and @ShivangiReja:  given that you're currently working in the `EventProcessorClient` area, your insight would be valuable.  Would you both please be so kind as to review?

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs]: nit:  Indentation is off after the change.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/EventProcessorClient/EventProcessorClientTests.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/EventProcessorClient/EventProcessorClientTests.cs]: Orthogonal to these changes, but the bracket spacing is off.    Also, AMGP is my favorite protocol!     

//cc: @kinelski.   Either you or I are guilty here.  😄

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/MockEventProcessorClient.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/MockEventProcessorClient.cs]: This is a really nice approach!

**Review by jsquire — COMMENTED** [review]
@christothes:  Thank you, again, for your contribution.   The approach looks great.  I wanted to let you know that there's a late-breaking tweak to the design that we're about to take, which is likely to cause a non-trivial bit of churn, unfortunately. 

I think the approach is remains relevant and valid.   I'd love to see us work with you as things take shape over the next week or so to incorporate these changes, if you're willing.  

If you'd prefer not, that's understandable and we remain grateful for your efforts and will definitely take advantage of many of the things that you've done here.

//cc: @kinelski, @ShivangiReja

**Comment by christothes** [comment]
> @christothes: Thank you, again, for your contribution. The approach looks great. I wanted to let you know that there's a late-breaking tweak to the design that we're about to take, which is likely to cause a non-trivial bit of churn, unfortunately.
> 
> I think the approach is remains relevant and valid. I'd love to see us work with you as things take shape over the next week or so to incorporate these changes, if you're willing.
> 
> If you'd prefer not, that's understandable and we remain grateful for your efforts and will definitely take advantage of many of the things that you've done here.
> 
> //cc: @kinelski, @ShivangiReja

I'd be happy to work through the changes.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/MockEventProcessorClient.cs
Comment (christothes) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/MockEventProcessorClient.cs]: thanks for the feedback!

**Comment by jsquire** [comment]
Hi @christothes:

Thank you for bearing with us.  There was a lot of churn in this area that we needed to push through for the pending release.  The last of the structural and feature changes should now be merged, so we're back to stable.   We've got a couple of tasks that we need to close out for release over the next day or so, and then @kinelski and I will loop back around and start working through things to help get this incorporated.

Thank you, again, for your contribution and patience.

**Comment by christothes** [comment]
Thanks for the update @jsquire. I have a local version of this branch with the merge issues resolved. It just needs a bit more work on the mocking approach, which has changed quite a bit in response to the structural changes. I should be able to push something by end of week.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs]: nit:  If you would be so kind, could you line up the second and third parameter with the `partitionId` declaration?

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Infrastructure/EventProcessorManager.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Infrastructure/EventProcessorManager.cs]: nit:  If you would be so kind, could you line up `second` with the declaration for `first`?

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/EventProcessorClient/EventProcessorClientTests.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/EventProcessorClient/EventProcessorClientTests.cs]: If we're using these outside the context of this class, we should probably look to move them to a dedicated class or set of classes under `Infrastructure`.

**Review by jsquire — APPROVED** [review]
There were just a couple of small nits that I've asked about.   Otherwise, I think we're in a good place to move this forward.

**Comment by jsquire** [comment]
@kinelski - Please let me know if you have any concerns or requests.   Once we've closed on a couple of minor items, I'm planning on bringing this in.

**Comment by christothes** [comment]
I think many of the comments here will be addressed once #8816 goes in. I'll merge them up after that one is merged.

**Comment by jsquire** [comment]
> I think many of the comments here will be addressed once #8816 goes in. I'll merge them up after that one is merged.

Good deal.  I noticed a bunch of overlap, but I wasn't sure if you were still intending them as separate.   I'll stand-down on this one until the other closes out.

**Comment by christothes** [comment]
This will be fixed up for completion once #9198 is merged.

**Comment by christothes** [comment]
This PR is now mostly obsolete

**Pull request closed without merging**
